File: backends/node/services/geo/geo-client/src/handlers/q/get-contacts-by-ext-keys.ts
import { BaseHandler, type IHandlerContext, validators } from "@d2/handler";
import { D2Result } from "@d2/result";
import type { GeoServiceClient, GetContactsByExtKeysResponse, ContactDTO } from "@d2/protos";
import type { MemoryCacheStore } from "@d2/cache-memory";
import { z } from "zod";
import type { GeoClientOptions } from "../../geo-client-options.js";
import { Queries } from "../../interfaces/index.js";

type Input = Queries.GetContactsByExtKeysInput;
type Output = Queries.GetContactsByExtKeysOutput;

/**
 * Handler for fetching Geo contacts by ext keys with local cache-aside.
 * Contacts are immutable, so cached entries never expire (only LRU eviction).
 * Fail-open: returns whatever was cached if gRPC fails.
 * Mirrors D2.Geo.Client.CQRS.Handlers.Q.GetContactsByExtKeys in .NET.
 */
export class GetContactsByExtKeys
  extends BaseHandler<Input, Output>
  implements Queries.IGetContactsByExtKeysHandler
{
  override get redaction() {
    return Queries.GET_CONTACTS_BY_EXT_KEYS_REDACTION;
  }

  private readonly store: MemoryCacheStore;
  private readonly geoClient: GeoServiceClient;
  private readonly inputSchema: z.ZodType<Input>;

  constructor(
    store: MemoryCacheStore,
    geoClient: GeoServiceClient,
    options: GeoClientOptions,
    context: IHandlerContext,
  ) {
    super(context);
    this.store = store;
    this.geoClient = geoClient;
    this.inputSchema = z
      .object({
        keys: z.array(
          z
            .object({ contextKey: validators.zodAllowedContextKey(options.allowedContextKeys) })
            .passthrough(),
        ),
      })
      .passthrough() as unknown as z.ZodType<Input>;
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    const validation = this.validateInput(this.inputSchema, input);
    if (!validation.success) return D2Result.bubbleFail(validation);

    if (input.keys.length === 0) {
      return D2Result.ok({ data: { data: new Map() }, traceId: this.traceId });
    }

    const result = new Map<string, ContactDTO[]>();
    const missingKeys: Array<{ contextKey: string; relatedEntityId: string }> = [];

    // Check cache first
    for (const key of input.keys) {
      const cacheKey = `contact-ext:${key.contextKey}:${key.relatedEntityId}`;
      const cached = this.store.get<ContactDTO[]>(cacheKey);
      if (cached !== undefined) {
        result.set(`${key.contextKey}:${key.relatedEntityId}`, cached);
      } else {
        missingKeys.push(key);
      }
    }

    // All cached — return early
    if (missingKeys.length === 0) {
      return D2Result.ok({ data: { data: result }, traceId: this.traceId });
    }

    // Fetch cache misses from Geo service
    let response: GetContactsByExtKeysResponse;
    try {
      response = await new Promise<GetContactsByExtKeysResponse>((resolve, reject) => {
        this.geoClient.getContactsByExtKeys({ keys: missingKeys }, (err, res) => {
          if (err) reject(err);
          else resolve(res);
        });
      });
    } catch {
      // Fail-open: return whatever was cached
      this.context.logger.warn(
        `gRPC call to Geo service failed for GetContactsByExtKeys. TraceId: ${this.traceId}`,
      );
      return D2Result.ok({ data: { data: result }, traceId: this.traceId });
    }

    if (response.result?.success && response.data) {
      // Cache each fetched result (no TTL — contacts are immutable, LRU evicts)
      for (const entry of response.data) {
        if (entry.key) {
          const mapKey = `${entry.key.contextKey}:${entry.key.relatedEntityId}`;
          const cacheKey = `contact-ext:${entry.key.contextKey}:${entry.key.relatedEntityId}`;
          this.store.set(cacheKey, entry.contacts);
          result.set(mapKey, entry.contacts);
        }
      }
    }

    return D2Result.ok({ data: { data: result }, traceId: this.traceId });
  }
}

export type {
  GetContactsByExtKeysInput,
  GetContactsByExtKeysOutput,
} from "../../interfaces/q/get-contacts-by-ext-keys.js";

File: backends/dotnet/services/Geo/Geo.Client/CQRS/Handlers/Q/GetContactsByExtKeys.cs
// -----------------------------------------------------------------------
// <copyright file="GetContactsByExtKeys.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Geo.Client.CQRS.Handlers.Q;

using D2.Services.Protos.Geo.V1;
using D2.Shared.Handler;
using D2.Shared.Interfaces.Caching.InMemory.Handlers.R;
using D2.Shared.Interfaces.Caching.InMemory.Handlers.U;
using D2.Shared.Result;
using FluentValidation;
using Grpc.Core;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using H = D2.Geo.Client.Interfaces.CQRS.Handlers.Q.IQueries.IGetContactsByExtKeysHandler;
using I = D2.Geo.Client.Interfaces.CQRS.Handlers.Q.IQueries.GetContactsByExtKeysInput;
using O = D2.Geo.Client.Interfaces.CQRS.Handlers.Q.IQueries.GetContactsByExtKeysOutput;

/// <summary>
/// Handler for fetching Geo contacts by ext keys with local cache-aside.
/// </summary>
/// <remarks>
/// Contacts are immutable, so cached entries never expire (only LRU eviction).
/// Fail-open: returns whatever was cached if gRPC fails.
/// </remarks>
public class GetContactsByExtKeys : BaseHandler<GetContactsByExtKeys, I, O>, H
{
    private readonly IRead.IGetHandler<List<ContactDTO>> r_cacheGet;
    private readonly IUpdate.ISetHandler<List<ContactDTO>> r_cacheSet;
    private readonly GeoService.GeoServiceClient r_geoClient;
    private readonly GeoClientOptions r_options;

    /// <summary>
    /// Initializes a new instance of the <see cref="GetContactsByExtKeys"/> class.
    /// </summary>
    /// <param name="cacheGet">The in-memory cache get handler for reading cached contacts.</param>
    /// <param name="cacheSet">The in-memory cache set handler for storing fetched contacts.</param>
    /// <param name="geoClient">The gRPC client for the Geo service.</param>
    /// <param name="options">The Geo client configuration options.</param>
    /// <param name="context">The handler context providing logging and tracing.</param>
    public GetContactsByExtKeys(
        IRead.IGetHandler<List<ContactDTO>> cacheGet,
        IUpdate.ISetHandler<List<ContactDTO>> cacheSet,
        GeoService.GeoServiceClient geoClient,
        IOptions<GeoClientOptions> options,
        IHandlerContext context)
        : base(context)
    {
        r_cacheGet = cacheGet;
        r_cacheSet = cacheSet;
        r_geoClient = geoClient;
        r_options = options.Value;
    }

    /// <inheritdoc />
    protected override HandlerOptions DefaultOptions => new(LogOutput: false);

    /// <inheritdoc />
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        var v = await ValidateInput(
            v => v.RuleForEach(i => i.Keys)
                .ChildRules(cr => cr.RuleFor(k => k.ContextKey)
                    .IsAllowedContextKey(r_options.AllowedContextKeys)),
            input,
            ct);
        if (v.Failed)
        {
            return D2Result<O?>.BubbleFail(v);
        }

        if (input.Keys.Count == 0)
        {
            return D2Result<O?>.Ok(new O([]));
        }

        var result = new Dictionary<string, List<ContactDTO>>();
        var missingKeys = new List<GetContactsExtKeys>();

        // Check cache first.
        foreach (var key in input.Keys)
        {
            var cacheKey = $"contact-ext:{key.ContextKey}:{key.RelatedEntityId}";
            var getR = await r_cacheGet.HandleAsync(new(cacheKey), ct);
            if (getR.CheckSuccess(out var cached) && cached?.Value is not null)
            {
                result[$"{key.ContextKey}:{key.RelatedEntityId}"] = cached.Value;
            }
            else
            {
                missingKeys.Add(key);
            }
        }

        // All cached — return early.
        if (missingKeys.Count == 0)
        {
            return D2Result<O?>.Ok(new O(result));
        }

        // Fetch cache misses from Geo service.
        GetContactsByExtKeysResponse response;
        try
        {
            response = await r_geoClient.GetContactsByExtKeysAsync(
                new GetContactsByExtKeysRequest { Keys = { missingKeys } },
                cancellationToken: ct);
        }
        catch (RpcException ex)
        {
            // Fail-open: return whatever was cached.
            Context.Logger.LogWarning(
                ex,
                "gRPC call to Geo service failed for GetContactsByExtKeys. TraceId: {TraceId}",
                TraceId);

            return D2Result<O?>.Ok(new O(result));
        }

        if (response.Result is { Success: true })
        {
            // Cache each fetched result (no expiration — contacts are immutable, LRU evicts).
            foreach (var entry in response.Data)
            {
                if (entry.Key is not null)
                {
                    var mapKey = $"{entry.Key.ContextKey}:{entry.Key.RelatedEntityId}";
                    var cacheKey = $"contact-ext:{entry.Key.ContextKey}:{entry.Key.RelatedEntityId}";
                    var contacts = entry.Contacts.ToList();

                    var setR = await r_cacheSet.HandleAsync(new(cacheKey, contacts), ct);
                    if (setR.Failed)
                    {
                        Context.Logger.LogWarning(
                            "Failed to cache contact-ext:{ContextKey}:{RelatedEntityId}. TraceId: {TraceId}",
                            entry.Key.ContextKey,
                            entry.Key.RelatedEntityId,
                            TraceId);
                    }

                    result[mapKey] = contacts;
                }
            }
        }

        return D2Result<O?>.Ok(new O(result));
    }
}
